#### no-trailing-spaces

Disallow trailing whitespace at the end of lines. Similar to https://eslint.org/docs/rules/no-trailing-spaces.

Bad:

```hbs
<div>test</div>//••
//•••••
```

Good:

```hbs
<div>test</div>//
//
```
